devito/operator.py

```diff
@@ -133,7 +134,9 @@ def expr_cse(expr):
     new_stencils = []
 
     for stencil in stencils:
-        new_stencils.append(stencil)
+        new_stencils.append(Eq(stencil.lhs,
+                               collect(stencil.rhs,
+                                       [temp.lhs for temp in to_keep])))
```

Comment:
OK, Two things:
1) this is not the right place to do these sort of things. CSE (semantic of the function) has nothing to do with factorization -- they are completely different transformations
2) as you're aware, i'm working on this symbolic engine to handle factorization automatically "behind the scenes", which will cope well with acoustic, tti and potentially many other problems. I'd encourage to remove this bit (perhaps you can cherry-pick the commit onto a temporary branch) and simply wait for this functionality, which will land with my next PR, hopefully by the end of the week